sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+
+A principal value represents either a set of keys, or a smart contract.
+The former, called a _standard principal_,
+is encoded as a `(buff 1)` *version byte*, indicating the type of account
+and the type of network that this principal can spend tokens on,
+and a `(buff 20)` *public key hash*, characterizing the principal's unique identity.
+The latter, a _contract principal_, is encoded as a standard principal concatenated with
+a `(string-ascii 40)` *contract name* that identifies the code body.
+
+`principal-destruct` will decompose a principal into its component parts.  Standard principals will 
+be decomposed into a tuple containing the **version byte** and **public key
+hash**.  Decomposed contract principals contain the same data as standard
+principals, as well as a **name** element that contains the human-readable
+contract name part of the contract address.
+
+This method returns a `Response` that wraps this data as a tuple.
+
+If the version byte of `principal-address` matches the network (see `is-standard`), then this method
+returns the pair as its `ok` value.
+
+If the version byte of `principal-address` does not match the network, then this method
+returns the pair as its `err` value.
+
+In both cases, the value itself is a tuple containing three fields: a `version` value as a `(buff 1)`,
+a `hash-bytes` value as a `(buff 20)`, and a `name` value as an `(optional (string-ascii 40))`.  The `name`
+field will only be `(some ..)` if the principal is a contract principal.
+
+**Rationale:** This method is meant to help smart contracts integrate with
+Bitcoin addresses and with other Stacks network instances' addresses.  Legacy
+Bitcoin addresses can already be converted to Stacks addresses; the ability to
+decode them in Clarity empowers developers to do things like verify that a
+Bitcoin owner signed some data, or also controlled some Stacks data.
+
+**Examples:**
+
+```clarity
+(principal-destruct 'STB44HYPYAT2BB2QE513NSP81HTMYWBJP02HPGK6) ;; Returns (ok (tuple (hash-bytes 0x164247d6f2b425ac5771423ae6c80c754f7172b0) (name none) (version 0x1a)))
+(principal-destruct 'STB44HYPYAT2BB2QE513NSP81HTMYWBJP02HPGK6.foo) ;; Returns (ok (tuple (hash-bytes 0x164247d6f2b425ac5771423ae6c80c754f7172b0) (name (some "foo")) (version 0x1a)))
+(principal-destruct 'SP3X6QWWETNBZWGBK6DRGTR1KX50S74D3433WDGJY) ;; Returns (err (tuple (hash-bytes 0xfa6bf38ed557fe417333710d6033e9419391a320) (name none) (version 0x16)))
+(principal-destruct 'SP3X6QWWETNBZWGBK6DRGTR1KX50S74D3433WDGJY.foo) ;; Returns (err (tuple (hash-bytes 0xfa6bf38ed557fe417333710d6033e9419391a320) (name (some "foo")) (version 0x16)))
+```
+
+### New method: `principal-construct`
+
+* **Input Signature:** `(principal-construct (buff 1) (buff 20) [(string-ascii 40)])`,
+* **Output Signature:** `(response principal { error_code: uint, value: (optional principal) })`
```

Comment:
Same as  with `principal-destruct` we have one function that does two things.
And again I will recommend spiting it into two separate functions:

#### principal-construct
* **Input Signature*: `(principal-construct (buff 1) (buff 20) (optional (string-ascii 40))`,
* **Output Signature**: `(response principal uint)`

#### is-network-principal
* **Input Signature**: `(is-network-principal (principal-address principal))`
* **Output Signature**: `bool`

Reply:
See https://github.com/stacksgov/sips/pull/95#discussion_r1007252711